chapter.md

```diff
@@ -0,0 +1,735 @@
+how those pieces interact,
+and how that decomposition enables and constrains further development.
+
+In this chapter,
+we will begin our exploration of the architecture of JavaScript applications
+by looking at how to use ideas borrowed from pure functional programming
+to simplify one of the most common tasks in web development:
+form handling.
+At first glance these two topics have nothing to do with each other,
+but paradoxically,
+acting as if the state of our system couldn't be changed
+actually makes its changes easier to manage.
+
+## The Problem
+
+Users can interact with programs in a bewildering variety of ways,
+but the most popular on the web remains the HTML form.
+In a classic 1990s-style application,
+a user is presented with a form,
+fills in the fields,
+and clicks "Submit".
+This sends an HTTP POST request to a server
+with the information the user entered as its payload.
+The server validates that information,
+updates a database,
+and returns an HTML page that may or may not contain another form.
+
+As browsers have become more powerful,
+and as interaction has moved away from the desktop to mobile devices,
+developers have moved more and more of this processing cycle to the client.
+The first step was to write a little JavaScript to sanity-check the data the user entered
+before POSTing it to the server.
+Programmers quickly began using JavaScript to make form interactions more fluid
+by showing and hiding sections based on what users had entered so far,
+and by fetching bits of data to do things like auto-completion while users were still typing.
+
+All of this made for a better user experience,
+but at the price of increasing client-side complexity.
+A modern single-page application (SPA) may have to keep track of dozens of pieces of information,
+any of which may need to be updated based on the user's actions
+*and* kept in sync with permanent storage on the server.
+As if that wasn't complicated enough,
+many of these interactions have to be done asynchronously
+in order to keep the user's FQ (frustration quotient) down.
+If,
+for example,
+a web page locks up for half a second every time the user types a single character
+because it's fetching possible auto-completes from the server,
+the user will quickly take her business somewhere else.
+
+Our problem is therefore this:
+
+> How should we manage client-side state in an asynchronous web application that uses forms?
+
+## Enter Redux
+
+Luckily for us,
+people who use pure functional programming languages
+have been thinking about these issues for more than thirty years,
+and have developed some design patterns that we can use
+in conventional languages like JavaScript.
+A *pure functional* language is one in which data cannot be changed,
+or *mutated*,
+in place:
+once a value is defined,
+it is *immutable*.
+Rather than changing its state,
+a program written in a pure functional language creates an entirely new state
+on which to operate.
+
+This may seem wasteful:
```

Comment:
what do you think about mentioning shallow copies (cheapest copy ever)? enforcing that it's not that bad since just the piece being changed is a new object everything else is the same reference